File: src/Algorithms/Sorting/Bubble Sort/code.py
from AlgorithmVisualizer import Tracer, Array1DTracer, LogTracer, Layout, VerticalLayout, ChartTracer, randomize

# define tracer variables {
chart = ChartTracer("Bubble Sort - Graph")
tracer = Array1DTracer("Bubble Sort - Array")
logger = LogTracer()
Layout.setRoot(VerticalLayout([chart, tracer, logger]))
array = randomize.Array1D(15).array
tracer.set(array)
tracer.chart(chart)
Tracer.delay()
# }

string_array = [str(int) for int in array]
println('original array = ['+','.join(string_array)+']')

swapped=True
n=len(array)
while(swapped):
	swapped=False
	# visualize {
	tracer.select(n-1)
	Tracer.delay()
	# }
	for i in range(1,n):
		# visualize {
		tracer.select(i-1,i)
		Tracer.delay()
		# }
		if(array[i-1]>array[i]):
			println('swap '+str(array[i - 1])+ ' and '+ str(array[i]))
			array[i - 1], array[i] = array[i], array[i - 1]
			swapped=True
			# visualize {
			tracer.patch(i - 1, array[i - 1])
			tracer.patch(i, array[i])
			Tracer.delay()
			tracer.depatch(i - 1)
			tracer.depatch(i)
			# }
		# visualize {
		tracer.deselect(i-1,i)
		# }
	# visualize {
	tracer.deselect(n-1)
	tracer.selectTrue(n-1)
	# }
	if not swapped:
		# visualize {
		tracer.selectTrue(0,n - 1)
		# }
	n-=1

string_array = [str(int) for int in array]
println('sorted array = ['+','.join(string_array)+']')

File: src/Algorithms/Sorting/Bubble Sort/code.js
// import visualization libraries {
const {
  Tracer,
  Array1DTracer,
  ChartTracer,
  LogTracer,
  Randomize,
  Layout,
  VerticalLayout,
} = require("algorithm-visualizer");
// }

// define tracer variables {
const chart = new ChartTracer("Bubble Sort - Graph ");
const tracer = new Array1DTracer("Bubble Sort - Array ");
const logger = new LogTracer();
Layout.setRoot(new VerticalLayout([chart, tracer, logger]));
const D = Randomize.Array1D({ N: 15 });
tracer.set(D);
tracer.chart(chart);
Tracer.delay();
// }

// logger {
logger.println(`original array = [${D.join(", ")}]`);
// }
let N = D.length;
let swapped;
do {
  swapped = false;
  // visualize {
  tracer.select(N - 1);
  Tracer.delay();
  // }
  for (let i = 1; i < N; i++) {
    // visualize {
    tracer.select(i-1,i);
    Tracer.delay();
    // }
    if (D[i - 1] > D[i]) {
      // logger {
      logger.println(`swap ${D[i - 1]} and ${D[i]}`);
      // }
      const temp = D[i - 1];
      D[i - 1] = D[i];
      D[i] = temp;
      swapped = true;
      // visualize {
      tracer.patch(i - 1, D[i - 1]);
      tracer.patch(i, D[i]);
      Tracer.delay();
      tracer.depatch(i - 1);
      tracer.depatch(i);
      // }
    }
    // visualize {
    tracer.deselect(i-1,i);
    // }
  }
  // visualize {
  tracer.deselect(N - 1);
  tracer.selectTrue(N - 1);
  if(!swapped){
    tracer.selectTrue(0,N - 1);
  }
  // }
  N--;
} while (swapped);
// logger {
logger.println(`sorted array = [${D.join(", ")}]`);
// }
